### ISSUE TITLE
[WoocommerceFluxstoreAPP] - Dropdown list for City Selection is not working

### ISSUE THREAD

**Issue by lemarchexpress** [inspireui/support#2672]
I would like to include a Dropdownlist for City selection & ZipCode Selection in the Checkout form (Introducing the shipping Address).

Regarding State Province you have included this solution so I have just copy and adapt the script.
However is not working, Could you tell me what is missing? This are the actions I have done for the city case:

I have created /cities/city_es.json file to call and list the Cities inside by country

![City_dropdownlist](https://user-images.githubusercontent.com/63209110/79068585-56377580-7cc8-11ea-805a-871b57b9b3da.png)

Here you can find attached my city_es.json, tools.dart and shipping_address.dart files...
[city_dropdownlist.zip](https://github.com/inspireui/support/files/4466286/city_dropdownlist.zip)

**Include the declaration of the new variable in lib/screens/checkout/shipping_address.dart:** (line 40)
  List<dynamic> cities = [];

**Include this widget in lib/screens/checkout/shipping_address.dart:** (Line: 564)
  Widget renderCityInput() {
    if (cities.isNotEmpty) {
      List<DropdownMenuItem> items = [];
      cities.forEach((item) {
        items.add(
          DropdownMenuItem(
            child: Text(item is Map ? item["name"] : item),
            value: item is Map ? item["code"] : item,
          ),
        );
      });
      String value;
      if (cities[0] is Map) {
        if (cities.firstWhere((o) => o["code"] == address.city,
            orElse: () => null) !=
            null) {
          value = address.city;
        }
      } else {
        value = cities.contains(address.city) ? address.city : null;
      }
      return DropdownButton(
        items: items,
        value: value,
        onChanged: (val) {
          setState(() {
            address.city = val;
          });
        },
        isExpanded: true,
        itemHeight: 70,
        hint: Text("Ciudad"),
      );
    } else {
      return TextFormField(
        controller: _cityController,
        validator: (val) {
          return val.isEmpty ? S.of(context).cityIsRequired: null;
        },
        decoration: InputDecoration(labelText: S.of(context).city),
        onSaved: (String value) {
          address.city = value;
        },
      );
    }
  }

**Include the modification of the widget (form) in lib/screens/checkout/shipping_address.dart:** (line 394)
                 
              renderCityInput(),

                    //TextFormField(
                      //controller: _cityController,
                      //validator: (val) {
                        //return val.isEmpty
                          //  ? S.of(context).cityIsRequired
                            //: null;
                      //},
                      //decoration:
                        //  InputDecoration(labelText: S.of(context).city),
                      //onSaved: (String value) {
                        //address.city = value;
                      //},
                    //),

**Include the modification of initstate in lib/screens/checkout/shipping_address.dart:** (line 74)
        cities = await Tools.loadCitiesByCountry(address.country);
        setState(() {});

**Include the modification of the form in lib/screens/checkout/shipping_address.dart:** (line 286)
                                    cities = await Tools.loadCitiesByCountry(address.country);
                                    setState(() {});

**Include the modification in lib/common/tools.dart:** (line 288)

  static Future<List<dynamic>> loadCitiesByCountry(String country) async {
    try {
      // load local config
      String path = "lib/config/cities/city_${country.toLowerCase()}.json";
      final appJson = await rootBundle.loadString(path);
      return List<dynamic>.from(convert.jsonDecode(appJson));
    } catch (e) {
      return [];
    }
  }
}

**Comment by lemarchexpress** [comment]
![image](https://user-images.githubusercontent.com/63209110/79070011-f397a700-7cd2-11ea-912b-f096dcb4a999.png)
Proof of being a customer...

**Comment by minhcasi** [comment]
@lemarchexpress - thank you for your all feedback, please help to create one ticket for all the issues

**Comment by minhcasi** [comment]
@philipvu30 - please help to investigate these issues include the closed issue above - thank you!

**Comment by vumanhsn97** [comment]
- shipping method: Replace this line in shipping_method model https://tppr.me/6CWUd
- stripe error: Please update mstore api
- logic for blocking order: Add this code in lib/screens/checkout/review.dart https://tppr.me/Q06Ak
- paypal: need to correct account
- customize left menu side: Replace this code in lib/menu.dart https://tppr.me/oAi1s
- update profile: working well. Email need to correct
- help same title: Replace file in lib/screens/search/search.dat to 
[search.dart.zip](https://github.com/inspireui/support/files/4468303/search.dart.zip) an lib/screens/cart/my_cart.dart to 
[my_cart.dart.zip](https://github.com/inspireui/support/files/4468305/my_cart.dart.zip)
- Tabbar: not found you issue: https://tppr.me/bhFAm

**Comment by lemarchexpress** [comment]
1) shipping method: Replace this line in shipping_method model https://tppr.me/6CWUd
**Feedback:** ISSUE SOLVED!

2) stripe error: Please update mstore api
**Feedback:** After updating the problem persists, the message returned is the same. Until now i have no payment method active working on the app. 

3) logic for blocking order: Add this code in lib/screens/checkout/review.dart https://tppr.me/Q06Ak
**Feedback:** Perfect, ISSUE SOLVED!

4) paypal: need to correct account
**Feedback:** I have used 3 different confirmed accounts and the problem still there. I cannot login and there is not payment method available on my app. 

5) customize left menu side: Replace this code in lib/menu.dart https://tppr.me/oAi1s
**Feedback:**  ISSUE SOLVED!

6) update profile: working well. Email need to correct
**Feedback:** The problem still there. it responds in the same way as reported.

7) help same title: Replace file in lib/screens/search/search.dat to
search.dart.zip an lib/screens/cart/my_cart.dart to
my_cart.dart.zip
**Feedback:**  
Regarding my cart: you have sent me a .dart file for Magento and my product is woocommerce. Would you my to send me Woocommerce version?
Regarding

8)Tabbar: not found you issue: https://tppr.me/bhFAm
**Feedback:**  I will double check the data..

9) REPLY ABOUT THIS TICKET IS STILL PENDING...
 [WoocommerceFluxstoreAPP] - Dropdown list for City Selection is not working #2672

10) NEW PROBLEM
After changes My Cart Summary is looking in this way...
![image](https://user-images.githubusercontent.com/63209110/79103032-2b9ef880-7d6c-11ea-9ab6-369b182634bb.png)

**Comment by lemarchexpress** [comment]
NEW ISSUE
11. Historical Order. 
I have Customized Orders States that are not considered in the Status.
![image](https://user-images.githubusercontent.com/63209110/79114851-a411b380-7d84-11ea-8ffc-45ab63c3a7b0.png)

The flow in the happy path is as follow:
Pedido Nuevo (New PO) -> Procesando (Processing) -> En Reparto (Delivery) ->Completado (Completed)
We will not accept POs without immediate payment through Stripe or Paypal so the Status On-Hold and Pending Payment are not necessary.

New PO is starting state.
Processing requires and action from backend to change to next status.
Delivery requires and action from backend to change to next status.
Completed is closing state.

Requirement:
1. Refunds button must not be shown never is a backend action.
2. Cancel button can be clicked only once.
3. Cancel button cannot be clicked after Order is set as COMPLETED State.
4. Enabled status can be read from the Analysis Panel -> Option Adjustments
5. Status that demands an action can be read from the Analysis Panel -> Option Adjustments

![image](https://user-images.githubusercontent.com/63209110/79114197-f520a800-7d82-11ea-9c7a-f742cd1b15c8.png)

**Comment by vumanhsn97** [comment]
- stripe error: Can you send your admin account to vumanhsn97@gmail.com
- paypal: Create a account in https://www.sandbox.paypal.com/
- update profile: Try to create a new account and update this account.
- same tittle: This is correct code. Please checking by your self. May be linking to folder wrong because I use code in dev branch.
- dropdown city: Make sure you add link folder to assets in pubspec.yaml
- cart summary: Please check new code
- history order: We will check it

**Comment by lemarchexpress** [comment]
stripe error: Can you send your admin account to vumanhsn97@gmail.com            
**Feedback:**  I have created an Admin User with this mail vumanhsn97@gmail.comYou should receive a password in the mailbox.

paypal: Create a account in https://www.sandbox.paypal.com/            
**Feedback:**  This link is only for USA accounts. I am in Spain. Regional segmentation does not allow me to open an account there.my account is ready and confirmed. should be transparent.  
Stripe or Paypal are my critical path at this moment most of the things are configure but I cannot receive anypayments.   

same tittle: This is correct code. Please checking by your self. May be linking to folder wrong because I use code in dev branch.              
**Feedback:**  I will double check this.

update profile: Try to create a new account and update this account.            
**Feedback:**  Issue Solved.

Dropdown city: Make sure you add link folder to assets in pubspec.yaml              
**Feedback:**  Issue Solved.  

cart summary: Please check new code              
**Feedback:**  Issue Solved. 
 
history order: We will check it              
**Feedback:**  Thanks...

**Comment by lemarchexpress** [comment]
**User Profile Update** issue was not closed. Check this Video of the null reaction after clicking "Update" button. 
https://drive.google.com/file/d/1_jCuJ2a_Ruopnz15xC3ZR6MMXqVL2djT/view

**Comment by lemarchexpress** [comment]
Dears Minh / Vumanhsn97,
Do we have anyprogress about the 5 pending items still open?
This is pending backlog:

**1.- Payment Method is not working: Stripe (Fatal Error) >** CRITICAL for me
Issue description...
Stripe does not work it showns a FATAL ERROR
Also, paypal image is not shown and it is properly configured.
![image](https://user-images.githubusercontent.com/63209110/79426248-2d162e00-7fc3-11ea-86b0-1c40590f2079.png)

**2.- Payment Method is not working: Paypal (login) >** Major issue but Stripe solved it as a workaround.
Issue description...
Using a valid paypal account, it never login, so at this moment, either paypal or stripe as payment methods don't work. How can i fix this?
Is it possible to include an smart button from paypal so the payment can be done immediately?
![image](https://user-images.githubusercontent.com/63209110/79426203-17a10400-7fc3-11ea-8f68-22eef2b5abef.png)


**3.- Update the User Profile data: Check the video I have posted. > CRITICAL for me**
https://drive.google.com/file/d/1_jCuJ2a_Ruopnz15xC3ZR6MMXqVL2djT/view

**4.- Guide to include the Smartchat on different screens such as: Categories, Search and Static.**
Issue description...
Just need to let me know the code to be included in the screens to activate the floatting button as well as we can find it in the Settings Screen.

**5.- History Order Tracking: > CRITICAL for me** 
a) Logic behind Cancel Button 
b) Hide Refund Button 
c) Customized Status are not shown in the graphic.

Issue description...
I have Customized Orders States that are not considered in the Status.
![image](https://user-images.githubusercontent.com/63209110/79426056-e58fa200-7fc2-11ea-85d6-4e6ede00a2be.png)

The flow in the happy path is as follow:
Pedido Nuevo (New PO) -> Procesando (Processing) -> En Reparto (Delivery) ->Completado (Completed)
We will not accept POs without immediate payment through Stripe or Paypal so the Status On-Hold and Pending Payment are not necessary. After checking carefully the customer can change the status always despite is considered as completed the order. is not suitable for operations.

New PO is starting state.
Processing requires and action from backend to change to next status.
Delivery requires and action from backend to change to next status.
Completed is closing state.

Requirement:
Refunds button must not be shown never is a backend action.
Cancel button can be clicked only once.
Cancel button cannot be clicked after Order is set as COMPLETED State.
Enabled status can be read from the Analysis Panel -> Option Adjustments
Status that demands an action can be read from the Analysis Panel -> Option Adjustments
![image](https://user-images.githubusercontent.com/63209110/79424843-17a00480-7fc1-11ea-856e-36f94aac6908.png)

Kind Regards!

**Comment by minhcasi** [comment]
@lemarchexpress - we are working on next release to fix the payment issue - it will be available in 1-2 days - thanks

**Comment by lemarchexpress** [comment]
Thanks!

**Comment by lemarchexpress** [comment]
Dear Minh / Vu,
I think this is relevant and will dorp your product to the Top in Spanish Market.
**REDSYS is the main payment gateway in our country** (We normally only use Redsys or Paypal) https://pagosonline.redsys.es/desarrolladores.html
It **works similar as STRIPE** does, but it is focused on local banks in Spain.
Considering that you are working on it and my backend has the plugin, verify that it is compatible and you will be able to make it part of your portfolio.
**At this point, it is giving me the same error as STRIPE does. Check the attached picture.**
Kind regards,

![image](https://user-images.githubusercontent.com/63209110/79689155-42a18700-8253-11ea-975f-875adcba4b3d.png)

**Comment by minhcasi** [comment]
@lemarchexpress - could you try to make a test with latest fluxstore 1.7.0 and mstore-api

**Comment by lemarchexpress** [comment]
I will try. Thanks

**Comment by lemarchexpress** [comment]
We will open a new ticket for our only pending issue that we could not
solve. The integration of the Redsys Payment Gateway

**Comment by Jegufull** [comment]
@lemarchexpress Hola pudiste hacer la lista desplegable para seleccionar la ciudad en la app?

**Comment by lemarchexpress** [comment]
Se hace todo tal cual como lo he explicado en el primer post solo faltaba
hacer esto...

   - dropdown city: Make sure you add link folder to assets in pubspec.yaml

Esta probado en version 1.6.6b no se si en la nueva version funciona igual
